crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -526,26 +560,43 @@ impl RecursorDnsHandle {
         Ok(ns)
     }
 
+    /// Check if an IP address matches any networks listed in the configuration that should not be
+    /// sent recursive queries.
+    fn matches_do_not_query(&self, ip: IpAddr) -> bool {
+        match ip {
+            IpAddr::V4(ip) => self.do_not_query_v4.contains(&ip.into()),
+            IpAddr::V6(ip) => self.do_not_query_v6.contains(&ip.into()),
+        }
+    }
+
     #[cfg(feature = "dnssec")]
     pub(crate) fn record_cache(&self) -> &DnsLru {
         &self.record_cache
     }
-}
 
-async fn append_ips_from_lookup<E: fmt::Display, R: fmt::Debug, I: Iterator<Item = IpAddr>>(
-    extract_ips: impl Fn(R) -> I,
-    futures: &mut (impl Stream<Item = Result<R, E>> + Unpin),
-    config: &mut NameServerConfigGroup,
-    activity: &str,
-) {
-    while let Some(next) = futures.next().await {
-        match next {
-            Ok(response) => {
-                debug!("{activity} A or AAAA response: {response:?}");
-                config.append_ips(extract_ips(response), true);
-            }
-            Err(e) => {
-                warn!("{activity} resolution failed failed: {e}");
+    async fn append_ips_from_lookup<E: fmt::Display, R: fmt::Debug, I: Iterator<Item = IpAddr>>(
+        &self,
```

Comment:
It's good that this just works with self. Sometimes in the `async fn` impls I've found self being captured can cause pinning issues. looks like that's not a problem here.